### ISSUE TITLE
Tracking Issue for complex generic constants: `feature(generic_const_exprs)`

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#76560]
This is a tracking issue for complex generic expressions in constants which is still highly experimental.
The feature gate for the issue is `#![feature(generic_const_exprs)]`.

On stable all const generic arguments must be either:
- A fully concrete expression using no generic parameters, e.g. `Foo<{ bar(1 + 2) }>`
- A bare usage of a const parameter, e.g. `Foo<{ N }>`

This feature allows arbitrary usage of generic parameters in const generic arguments such that `Foo<{ bar(N + 2) }>` is now legal.

Initial proposal: https://github.com/rust-lang/compiler-team/issues/340
Design document: [HackMD document](https://hackmd.io/OZG_XiLFRs2Xmw5s39jRzA?view)

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Status

The design questions and implementation challenges are collected in the [project-const-generics repository](https://github.com/rust-lang/project-const-generics/issues?q=is%3Aissue+is%3Aopen+label%3AC-design-docs+label%3AA-generic-exprs). The implementation is still far from ready but already available for experimentation.

**Comment by newpavlov** [comment]
We had a short discussion relevant to this feature in #79778.

@oli-obk 
If we are to continue it, I think it will be better to do it here.

**Comment by oli-obk** [comment]
This issue is a tracking issue about the `const_evaluatable_checked` feature, any large scale changes of the kind we talked about should have their own issues (and refer to this issue) or probably better an internals thread, as github issues are not great for long running discussions.

**Comment by shepmaster** [comment]
> * fix bogus error on missing associated items

This may have been incidentally fixed.

**Comment by RalfJung** [comment]
The MCP mentions that

> arbitrary function calls

are allowed and will be "unified". Isn't that incompatible with impure `const fn` such as what one will be able to write once `const fn` can perform heap allocations?

**Comment by RalfJung** [comment]
Also, in terms of arithmetic, it might be better to exclude floating-point arithmetic since at least some models propose to make that arithmetic non-deterministic (specifically, the NaN payload may be non-deterministic) -- also see https://github.com/rust-lang/unsafe-code-guidelines/issues/237.

**Comment by lcnr** [comment]
> The MCP mentions that
> 
> > arbitrary function calls
> 
> are allowed and will be "unified". Isn't that incompatible with impure `const fn` such as what one will be able to write once `const fn` can perform heap allocations?

uh, now that's a challenge I didn't think of :laughing: not yet sure

how do we expect to deal with `[nondet_or_unpure_const_fn(); 4]`?
will we rerun the the const fn 4 times?

**Comment by RalfJung** [comment]
`[nondet_or_unpure_const_fn(); 4]` only compiles if the return type is `Copy`; it will call the function once and copy the result 4 times. No CTFE is involved.

The interesting case is `[const { nondet_or_unpure_const_fn() }; 4]`. I expect it evaluates the function once in CTFE and uses the result 4 times.

Note that the non-determinism is very restricted; it only arises in the `AllocId` of pointers. So maybe for the type system this simply does not matter (assuming "value trees" consist of integers only, no raw pointers). However, we have have some ptr comparison functions... they are rather weak but still, I wonder what one could do there.

**Comment by clarfonthey** [comment]
So I was looking through the current design for this feature and noticed that there isn't a clear definition of how const generics should interact with associated consts. I filed #80976 and #80977 for specific errors that show up when associated consts are used in const generics directly, but noticed that we don't actually have any defined syntax for how you would constrain associated consts in bounds, e.g. something like `T: Trait<const CONST = X>`.

I figure that this is probably something that needs to be figured out before this feature can be merged, even if the decision is that associated consts simply aren't allowed for MVP (which IMHO wouldn't be that great of an idea).

**Comment by linclelinkpart5** [comment]
@clarfonthey This is a feature I'm in mighty need of in my personal project, I'd love to be able to use associated consts more heavily and do something like this:
```rust

trait Sample: Copy + Clone + PartialOrd + PartialEq {
    const EQUILIBRIUM: Self;
}

trait Frame: Copy + Clone + PartialEq {
    type Sample: Sample;

    const NUM_CHANNELS: usize;
    const EQUILIBRIUM: Self;

    // Returns a fixed array of of `Self::Sample`s of size `Self::NUM_CHANNELS`.
    fn to_array(self) -> [Self::Sample; { Self::NUM_CHANNELS }];
}

// All fixed-size arrays of samples are frames.
impl<S, const N: usize> Frame for [S; N]
where
    S: Sample,
{
    type Sample = S;

    const NUM_CHANNELS: usize = N;
    const EQUILIBRIUM: Self = [S::EQUILIBRIUM; Self::NUM_CHANNELS];

    // Already an array, just return `self`.
    // I would hope [S; N] and [S; Self::NUM_CHANNELS] are realized to be the same type!
    fn to_array(self) -> [Self::Sample; { Self::NUM_CHANNELS }] {
        self
    }
}

// Individual samples can be treated as frames that always have only 1 channel.
impl<S> Frame for S
where
    S: Sample,
{
    type Sample = S;

    const NUM_CHANNELS: usize = 1;
    const EQUILIBRIUM: Self = S::EQUILIBRIUM;

    // Again, I would hope [S; 1] and [S; Self::NUM_CHANNELS] are realized to be the same type!
    fn to_array(self) -> [Self::Sample; { Self::NUM_CHANNELS }] {
        [self]
    }

// Standalone function to show example of const generic bound restrictions.
fn pair_test<A, B, F>(frame_a: A, frame_b: B, mut test_func: F) -> bool
where
    A: Frame,
    B: Frame<const NUM_CHANNELS = A::NUM_CHANNELS>, // Same constant number of channels.
    F: FnMut(&A::Sample, &B::Sample) -> bool,
{
    todo!("pretend this is a calculation that expects two arrays of the same size")
}
```

Right now however, using `#[min_const_generics]` on nightly, I have to resort to something like this:

```rust
trait Sample: Copy + Clone + PartialOrd + PartialEq {
    const EQUILIBRIUM: Self;
}

// This trait now has a const generic parameter.
trait Frame<const N: usize>: Copy + Clone + PartialEq {
    type Sample: Sample;

    const NUM_CHANNELS: usize;
    const EQUILIBRIUM: Self;

    // Returns a fixed array of of `Self::Sample`s of size `Self::NUM_CHANNELS`.
    fn to_array(self) -> [Self::Sample; N];
}

// All fixed-size arrays of samples are frames.
impl<S, const N: usize> Frame<N> for [S; N]
where
    S: Sample,
{
    type Sample = S;

    const NUM_CHANNELS: usize = N;
    const EQUILIBRIUM: Self = [S::EQUILIBRIUM; N];

    fn to_array(self) -> [Self::Sample; N] {
        self
    }
}

// Individual samples can be treated as frames that always have only 1 channel.
impl<S> Frame<1> for S
where
    S: Sample,
{
    type Sample = S;

    const NUM_CHANNELS: usize = 1;
    const EQUILIBRIUM: Self = S::EQUILIBRIUM;

    fn to_array(self) -> [Self::Sample; 1] {
        [self]
    }
}

// Now, I have to include a `const N: usize` bound, even though this new method
// doesn't directly use its value. The value of `N` is only used to ensure that
// the two frames have the same number of channels. This kind of const generic
// pollution is especially bad when returning values parameterized by types that
// have const generics.
fn pair_test<A, B, F, const N: usize>(frame_a: A, frame_b: B, mut test_func: F) -> bool
where
    A: Frame<N>,
    B: Frame<N>, // Same constant number of channels.
    F: FnMut(&A::Sample, &B::Sample) -> bool,
{
    todo!("pretend this is a calculation that expects two arrays of the same size")
}
```

**Comment by liarokapisv** [comment]
I am not sure if this is the proper issue to post this: Is the following expected to work? I would expect that the following deduction would work and so I wouldn't have to ad the extra bound on the create function:

```
[u8 ; size_assert::<HeapStorage, T>()] : Sized =>
[u8 ; HeapStorage::Asserter<T>::VALUE as usize - 1] : Sized =>
[u8 ; AlwaysTrueAsserter::VALUE as usize - 1] : Sized =>
[u8 ; true as usize - 1] : Sized =>
[u8 ; 0] : Sized

```
code:

```
#![no_std]

#![feature(generic_associated_types)]
#![feature(const_generics)]
#![feature(const_evaluatable_checked)]
#![feature(const_fn)]

#![allow(dead_code)]
#![allow(incomplete_features)]

use core::marker::PhantomData;
use core::mem::{size_of, align_of};

trait BoolConst {
    const VALUE : bool;
}

struct AlwaysTrueAsserter {}

impl BoolConst for AlwaysTrueAsserter {
    const VALUE : bool = true;
}

struct SizeAsserter<L,R>(PhantomData<(L,R)>);

impl<S,T> BoolConst for SizeAsserter<S, T> {
    const VALUE : bool = size_of::<S>() >= size_of::<T>() && 
                         align_of::<S>() % align_of::<T>() == 0;
}

trait Storage {
    type Asserter<T> : BoolConst;
}

struct HeapStorage {}

impl Storage for HeapStorage {
    type Asserter<T> = AlwaysTrueAsserter;
}

struct RawBox<T, S: Storage>(PhantomData<(T,S)>);

const fn size_assertion<S : Storage, T>() -> usize {
    S::Asserter::<T>::VALUE as usize - 1
}

impl<T, S : Storage> RawBox<T, S> {

    fn create(_val : T) -> Self where
        [u8 ; size_assertion::<S, T>()] : Sized
    {
        Self(PhantomData)
    }

}

fn create<T>(val : T) -> RawBox::<T, HeapStorage> 
    // this here is my issue
    where  [u8 ; size_assertion::<HeapStorage, T>()] : Sized
{
    RawBox::<T, HeapStorage>::create(val)
}
```

**Comment by clarfonthey** [comment]
I figure this isn't even something that would be considered during MVP, but is there any way we could potentially make:

```
#![feature(const_generics, const_evaluatable_checked)]
fn flatten<T, const N: usize, const M: usize>(arr: [[T; N]; M]) -> [T; N * M] {
    unsafe { core::mem::transmute(arr) }
}
```

work with the existing MVP? Right now `transmute` will always fail if two types' sizes fail to unify, rather than acting at monomorphisation time. Probably something to consider with existing design.

**Comment by RalfJung** [comment]
I don't think `transmute` will become smart like that. You can use `transmute_copy` though which has no size check (and is "even more unsafe" than `transmute`).

**Comment by lcnr** [comment]
> work with the existing MVP

yes, and while I might be missing something, I expect us to change transmute to use a `size_of::<T>() == size_of::<U>()` bound once this feature works well enough.

@RalfJung why do you expect this to not be possible?

**Comment by RalfJung** [comment]
Maybe I am underestimating the power of full-blown const generics.^^

**Comment by HTG-YT** [comment]
Probably a bump on this feature to be implemented.

(*Sorry if I somehow misunderstood the error, but I'm up for corrections*)
Currently I have a use case in which I want to call SIMD intrinsics on latest nightly (which indeed uses const generics). However, the code I am writing requires arithmetic on the consts. Here's my code:

```rust
    #[inline]
    #[target_feature(enable = "sse4.2")]
    unsafe fn prev<const IMM8: i32>(self, prev: Self) -> Self {
        Self::from(_mm_alignr_epi8::<{ 16 - IMM8 }>(
            self.0,
            prev.0
        ))
    }
```

Which currently does not compile with the following error:
```
error: unconstrained generic constant
   --> lightsaber_util\src\libcore_ext\str\validations\amd64\mod.rs:136:38
    |
136 |         Self::from(_mm_alignr_epi8::<{ 16 - IMM8 }>(
    |                                      ^^^^^^^^^^^^^
    |
    = help: try adding a `where` bound using this expression: `where [(); { 16 - IMM8 }]:`
```

I am not sure what does that error mean, but shouldn't that also be an "overly complex constant expression"? Thanks.

**Comment by mriise** [comment]
What prevents this?
```rust 
struct Foo<const N: u8> ([u8; N as usize]);
```
 Casting a u8 into a usize seems trivial, but compiler complains about "overly complex generic constant".

**Comment by clarfonthey** [comment]
It's because the constants don't currently unify. Technically you could do `N as u16 as usize` and it would always have the same value but those two constants would be considered separate at the type level.

The current code is designed to be as conservative as possible.

**Comment by mriise** [comment]
> 
> 
> It's because the constants don't currently unify.

Is there a specific tracking issue for this I could follow? I found similar issues but not for integer types...

or has this behavior been rejected?

**Comment by BoxyUwU** [comment]
I believe the reason this isnt supported is because we just haven't done the impl work yet, it should be possible to support

```rust
const fn my_cast(arg: u8) -> usize {
  arg as usize
}
struct Foo<const N: u8>([u8; my_cast(N)]);
```
^^^ this is a current workaround to cast stuff

**Comment by mriise** [comment]
I had tried this earlier but had forgotten to add back in the right feature flags!
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=6c45a92271c9ad26eb4f479b6728d266
Thanks!

**Comment by HTG-YT** [comment]
> 
> 
> Probably a bump on this feature to be implemented.
> 
> (_Sorry if I somehow misunderstood the error, but I'm up for corrections_)
> Currently I have a use case in which I want to call SIMD intrinsics on latest nightly (which indeed uses const generics). However, the code I am writing requires arithmetic on the consts. Here's my code:
> 
> ```rust
>     #[inline]
>     #[target_feature(enable = "sse4.2")]
>     unsafe fn prev<const IMM8: i32>(self, prev: Self) -> Self {
>         Self::from(_mm_alignr_epi8::<{ 16 - IMM8 }>(
>             self.0,
>             prev.0
>         ))
>     }
> ```
> 
> Which currently does not compile with the following error:
> 
> ```
> error: unconstrained generic constant
>    --> lightsaber_util\src\libcore_ext\str\validations\amd64\mod.rs:136:38
>     |
> 136 |         Self::from(_mm_alignr_epi8::<{ 16 - IMM8 }>(
>     |                                      ^^^^^^^^^^^^^
>     |
>     = help: try adding a `where` bound using this expression: `where [(); { 16 - IMM8 }]:`
> ```
> 
> I am not sure what does that error mean, but shouldn't that also be an "overly complex constant expression"? Thanks.

Are there any workarounds for this currently?

**Comment by pacak** [comment]
> Are there any workarounds for this currently?

This is stupid but seems to work.

```rust
⚠ const fn my_cast(arg: i32) -> usize {             
      arg as usize
  }        
                               
⚠ unsafe fn prev<const IMM8: i32, const SIXTEEN: i32>(a: [u8; 16], prev: [u8; 16]) -> [u8; 16]
  where                                                       
⚠     [(); my_cast({ SIXTEEN- IMM8 })]: ,           
  {                                           
      transmute(core::arch::x86_64::_mm_alignr_epi8::<{ SIXTEEN - IMM8 }>(
          transmute(a),                   
          transmute(prev),
      ))
  }                
```

**Comment by HTG-YT** [comment]
I tried it on the [Rust Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ca6292151968f7c4f194a894f857d3bb):
```
error: constant expression depends on a generic parameter
  --> src/lib.rs:12:5
   |
12 |     [(); my_cast({ SIXTEEN- IMM8 })]: ,           
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this may fail depending on what value the parameter takes
   ```

**Comment by pacak** [comment]
`#![feature(const_evaluatable_checked)]`

**Comment by HTG-YT** [comment]
Ahhh thanks. Alright

**Comment by jhpratt** [comment]
What's the current status of this, anyways? I see that it's still an incomplete feature — anything I might be able to assist with?

**Comment by Volker-Weissmann** [comment]
> I tried it on the [Rust Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ca6292151968f7c4f194a894f857d3bb):
> 
> ```
> error: constant expression depends on a generic parameter
>   --> src/lib.rs:12:5
>    |
> 12 |     [(); my_cast({ SIXTEEN- IMM8 })]: ,           
>    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>    |
>    = note: this may fail depending on what value the parameter takes
> ```

It would be nice if this code would produce
```
 = help: add `#![feature(const_evaluatable_checked)]` to the crate attributes to enable
```

**Comment by gmorenz** [comment]
In the following code, `const_main` fails to compile (apparently because it has a const generic parameter) while `main` (which has an identical body) compiles - i.e. commenting out `const_main` makes it work. I know the feature is incomplete but I found the behaviour surprising and thought it worth mentioning - also I believe it use to work (I haven't validated this exact code, but I'm trying to update an old project to a more recent nightly).

```rust
#![allow(incomplete_features)]
#![feature(
    const_generics,
    const_evaluatable_checked,
)]


/// Only works on powers of 2
pub const fn simple_log_2(x: usize) -> usize {
    x.trailing_zeros() as usize
}


fn foo<const N: usize>(arr: [usize; simple_log_2(N)]) {
    println!("{:?}", arr);
}

fn const_main<const M: usize>() {
    const N: usize = 2;
    foo::<N>([0])
}

fn main() {
    const N: usize = 2;
    foo::<N>([0])
}
```

Error

```
error: unconstrained generic constant
  --> src/main.rs:20:5
   |
20 |     foo::<N>([0])
   |     ^^^^^^^^
   |
   = help: try adding a `where` bound using this expression: `where [(); simple_log_2(N)]:`
note: required by a bound in `foo`
  --> src/main.rs:14:37
   |
14 | fn foo<const N: usize>(arr: [usize; simple_log_2(N)]) {
   |                                     ^^^^^^^^^^^^^^^ required by this bound in `foo`

error: could not compile `tmp` due to previous error
[Finished running. Exit status: 101]
```

**Comment by phaazon** [comment]
I’m currently trying this feature to make a change in `luminance` and just hit a compiler error. Not sure whether I should provide the snippet here, but in doubt, here it is.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1f3b71a20a79dee66d9a5e58c3d3b138

**Comment by memoryruins** [comment]
@phaazon for bugs like these, it's worth opening new issues rather than posting to tracking issues. this appears to be same ICE during type inference mentioned in #83249. if the types are provided in the `wrap2` calls in `main`

```diff
- let x = wrap2(1., [1., 2., 3.])
+ let x = wrap2::<f32, [f32; 3]>(1., [1., 2., 3.])
```

 (and the `const_fn_trait_bound` feature is enabled in your case), it compiles successfully.

**Comment by elichai** [comment]
Combining this with `adt_const_params` can give weird errors:
```rust
#![feature(generic_const_exprs, adt_const_params)]

struct FieldElement<const N: &'static str> {
    n: [u64; num_limbs(N)],
}
const fn num_limbs(_: &str) -> usize {
    0
}
```
Error:
```rust
error: overly complex generic constant
 --> src/lib.rs:5:14
  |
5 |     n: [u64; num_limbs(N)],
  |              ^^^^^^^^^^-^
  |                        |
  |                        dereferencing is not supported in generic constants
  |
  = help: consider moving this anonymous constant into a `const` function
  = note: this operation may be supported in the future
  ```
  https://play.rust-lang.org/?gist=a6a2aecfba83485968a81b8599b86cc5
  
  (I hoped I can try and play with these to see if it would be possible to implement typed algebraic fields)

**Comment by Logarithmus** [comment]
@lcnr what's the status of this issue? It looks to me that all PRs mentioned in issue's body are merged.

**Comment by clarfonthey** [comment]
Considering how the "fully implement this feature" box is still unchecked, I would hesitate to say that this feature is ready to be stabilised.

**Comment by Logarithmus** [comment]
@clarfonthey then where can I learn what steps are left to do?

**Comment by clarfonthey** [comment]
A lot, honestly. Right now, the hack of having no bounds on a type involving a constant is probably undesired, and I've seen a few different suggestions for how to do it. That said, there *are* a few cases that do not rely on this hack that could probably be stabilised, mostly involving associated consts. Since the existence of associated consts is always publicly exposed, we could easily assume that they're public APIs, and that stuff like `Trait<CONST = Type::CONST>` could be stabilised without needing to stabilise the full feature.

Already, I feel like allowing the use of associated consts in general could act as a workaround until the full feature is properly specked out, but I'm not on the lang team and have no idea what the const-wg has planned for this.

Perhaps someone from the lang team or the const-wg could help provide more insight? I've also seen a few blog posts regarding potential designs for this but definitely won't be able to find them easily, so, maybe if someone could share those as well it might provide some insight.

**Comment by tarcieri** [comment]
> Since the existence of associated consts is always publicly exposed, we could easily assume that they're public APIs, and that stuff like `Trait<CONST = Type::CONST>` could be stabilised without needing to stabilise the full feature.

Could something like that provide a solution for #60551? I think that's a major blocker in the usability of the current implementation of const generics.

**Comment by clarfonthey** [comment]
That issue is exactly the thing I was describing; thank you for linking it. Essentially, the idea would be to make #60551 work without adding empty bounds to where clauses, essentially reclassifying associated consts to work with the currently stabilised version of const generics.

**Comment by KisaragiEffective** [comment]
I think below code should be allowed in the future implementation:
```rs
#![feature(generic_const_exprs)]
struct T<const A: usize>([(); A]);

impl<const L: usize> T<L> {
    fn void(&self) -> T<0> {
        T([(); L - L])
    }
}
```

**Comment by mriise** [comment]
```rust
// stable rust
struct T<const A: usize> ([(); A]);

impl<const L: usize> T<L> {
    /// function holds for all T<L>
    fn void(&self) -> T<0> {
        T([(); 0])
    }
}
```

**Comment by lcnr** [comment]
The relevant issues for this are tracked in https://github.com/rust-lang/project-const-generics/issues?q=is%3Aissue+is%3Aopen+label%3AC-design-docs+label%3AA-generic-exprs

it's still going to take a while until we're close to stabilization here[^1], going to update the tracking issue though.

[^1]: to a large part because cg currently isn't the main focus of either myself or the compiler team

**Comment by psionic12** [comment]
Is there anyway I can do something like this in nightly?
```
pub struct Foo<const I: i32> {}

impl<const I: i32> Foo<I> where { I > 0 } {
  fn foo() -> SomeTypeA {}
}
impl<const I: i32> Foo<I> where { I == 0 } {
  fn foo() -> SomeTypeB {}
}
impl<const I: i32> Foo<I> where { I < 0 } {
  fn foo() -> SomeTypeC {}
}
```

**Comment by clarfonthey** [comment]
Nope, there's currently no way to bound const generics. The best you have right now is to change the type of the generic to affect the bound.

**Comment by VilNeo** [comment]
Hi,
one question: Will it be possible to support 'generic const expression' based types recursively?
So the structure I have in mind is something like a 'cons list' where each subsequent type has a const generic increased by one.

So what I want is something that will look like the following:

```
enum Cons<const N: usize>
where
    [(); N + 1]:,
{
    End,
    Next(Box<Cons<{ N + 1 }>>),
}
```

**Comment by oli-obk** [comment]
Unless we can start adding where bounds to variants, this will be an infinitely recursive type. So I think this is blocked on another feature that isn't even RFCed yet

**Comment by clarfonthey** [comment]
Yeah, boxing can prevent infinite size types, but it can't stop infinitely complex types, even if said types are finitely sized.

**Comment by VilNeo** [comment]
Yes, I had also seen this problem with infinite recursive type. And yes, one can probably omit the boxing here.

Is it a possibility here to apply conditional compilation as follows?

```rust
enum Cons<const N: usize>
where
    [(); N + 1]:,
{
    End,
    #[cfg(N < 20)]
    Next(Cons<{ N + 1 }>),
}

```

and with structs:
```rust
struct Cons<const N: usize>
where
    [(); N + 1]:,
{
    #[cfg(N < 20)]
    Option<Cons<{ N + 1 }>>,
}

```

I could imagine that conditional variants/fields would greatly increase the complexity when used later.

**Comment by oli-obk** [comment]
These kind of cfg are what I'd expect to happen as where bounds, but that's entirely orthogonal to const generics. Probably best to start a discussion on zulip or the internals forum, ideally with examples using types and traits instead of const generics (to reduce the number of feature gates needed to come up with examples)

**Comment by oxabz** [comment]
Hi,
Is a generic implementation of a trait possible (or planned) for a subset of a const generic?
Something  like this :  
```rs
#[derive(PartialEq, Eq, Debug)]
pub enum TensorSize{
    Static(usize),
    Dynamic
}

pub struct Tensor1<const U: TensorSize>;

impl<const S:usize> Add<Tensor1<{TensorSize::Static(S)}>> for Tensor1<{TensorSize::Static(S)}> {
    type Output = Tensor1<{TensorSize::Static(S)}>;
    //...
}

impl<const O:TensorSize> Add<Tensor1<O>> for Tensor1<{TensorSize::Dynamic}> {
    type Output = Tensor1<{TensorSize::Dynamic}>;
    //...
}
```

**Comment by KisaragiEffective** [comment]
@oxabz that is unlikely to able to have, because [RFC: enum variant types](https://github.com/rust-lang/rfcs/pull/2593) was postponed.

**Comment by psionic12** [comment]
I found something weird, not sure if it is a bug:
```
const ONE : i32 = 1;
const TWO : i32 = 2;

struct Foo<const i: i32, const j: i32> {}

impl<const i: i32> Foo<i, ONE> {
    pub fn foo() {}
}

impl<const i: i32> Foo<i, TWO> {
    pub fn foo() {}
}
```
This can comiple without `#![feature(generic_const_exprs)]`, but if `#![feature(generic_const_exprs)]` is added, it won't compile:
```
error[E0592]: duplicate definitions with name `foo`
  --> src/main.rs:9:5
   |
9  |     pub fn foo() {}
   |     ^^^^^^^^^^^^ duplicate definitions for `foo`
...
13 |     pub fn foo() {}
   |     ------------ other definition for `foo`
```

**Comment by coreylowman** [comment]
Is there a team working on this? Where can I find updates/progress/timeline/roadmap/etc?

**Comment by lcnr** [comment]
this is currently mostly tracked in https://github.com/rust-lang/project-const-generics/issues?q=is%3Aissue+is%3Aopen+label%3AA-generic-exprs and on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/260443-project-const-generics)

**Comment by schuelermine** [comment]
I’ve run into this case and I’m unsure if this *should* be figured out by the compiler. It definitely makes sense to me as a human…
```rs
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

struct ArrW<T, const N: usize>([T; N]);

trait IntoArrW<T> {
    const N: usize;
    fn into_arr_w(self) -> ArrW<T, { Self::N }>;
}

impl<T, const N: usize> IntoArrW<T> for [T; N] {
    const N: usize = N;
    fn into_arr_w(self) -> ArrW<T, { Self::N }> {
        ArrW(self)
    }
}
```
```
error[[E0308]](https://doc.rust-lang.org/nightly/error_codes/E0308.html): mismatched types
  --> src/lib.rs:14:14
   |
14 |         ArrW(self)
   |              ^^^^ expected `{ Self::N }`, found `N`
   |
   = note: expected constant `{ Self::N }`
              found constant `N`

For more information about this error, try `rustc --explain E0308`.
```

**Comment by KisaragiEffective** [comment]
I'd like to note for useful case.

Consider this trait:
```
pub trait AsBitPattern {
    /// Note: ZST must returns empty slice.
    fn as_bit_pattern(&self) -> &[u8];
}
```

and its blanket implementation:
```
impl<T> AsBitPattern for T {
    fn as_bit_pattern<'a>(&'a self) -> &'a [u8] {
        let start_addr = (self as *const T).cast::<u8>();
        // SAFETY:
        // The following property is always hold for Sized types:
        // - size_of::<T> == size_of::<T> * size_of::<u8>
        // - It will not truncate nor cause out-of-bound-read.
        let slice = unsafe {
            core::slice::from_raw_parts::<'a, u8>(start_addr, core::mem::size_of::<T>())
        };
        slice
    }
}
```

Using this feature, it can be rewritten as (currently, generics in const context is not allowed):
```
impl<T> AsBitPattern for T {
    fn as_bit_pattern<'a>(&'a self) -> &'a [u8] {
        // SAFETY:
        // The following property is always hold for Sized types:
        // - size_of::<T> == size_of::<T> * size_of::<u8>
        // ref-to-ref cast will not truncate nor cause out-of-bound-read.
        unsafe {
            core::mem::transmute::<&'a T, &'a [u8; core::mem::size_of::<T>()]>(self)
        }
    }
}
```

**Comment by Nugine** [comment]
> I'd like to note for useful case.
> 
> Consider this trait:
> 
> ```rust
> pub trait AsBitPattern {
>     /// Note: ZST must returns empty slice.
>     fn as_bit_pattern(&self) -> &[u8];
> }
> ```
> 
> and its blanket implementation:
> 
> ```rust
> impl<T> AsBitPattern for T {
>     fn as_bit_pattern<'a>(&'a self) -> &'a [u8] {
>         let start_addr = (self as *const T).cast::<u8>();
>         // SAFETY:
>         // The following property is always hold for Sized types:
>         // - size_of::<T> == size_of::<T> * size_of::<u8>
>         // - It will not truncate nor cause out-of-bound-read.
>         let slice = unsafe {
>             core::slice::from_raw_parts::<'a, u8>(start_addr, core::mem::size_of::<T>())
>         };
>         slice
>     }
> }
> ```
> 
> Using this feature, it can be rewritten as (currently, generics in const context is not allowed):
> 
> ```rust
> impl<T> AsBitPattern for T {
>     fn as_bit_pattern<'a>(&'a self) -> &'a [u8] {
>         // SAFETY:
>         // The following property is always hold for Sized types:
>         // - size_of::<T> == size_of::<T> * size_of::<u8>
>         // ref-to-ref cast will not truncate nor cause out-of-bound-read.
>         unsafe {
>             core::mem::transmute::<&'a T, &'a [u8; core::mem::size_of::<T>()]>(self)
>         }
>     }
> }
> ```

Note that this snippet is unsound. See the bound of [bytemuck::bytes_of](https://docs.rs/bytemuck/latest/bytemuck/fn.bytes_of.html).

**Comment by RalfJung** [comment]
You can also make it sound by using `MaybeUninit<u8>` instead of `u8`.

**Comment by programmerjake** [comment]
> You can also make it sound by using `MaybeUninit<u8>` instead of `u8`.

that's not actually true afaict, data race:

```rust
fn as_bit_pattern<T, 'a>(v: &'a T) -> &'a [MaybeUninit<u8>] {
    let start_addr = (v as *const T).cast::<MaybeUninit<u8>>();
    // SAFETY: not safe
    let slice = unsafe {
        core::slice::from_raw_parts(start_addr, core::mem::size_of::<T>())
    };
    slice
}

fn main() {
    static V: AtomicU8 = AtomicU8::new();
    let t = thread::spawn(|| V.store(1, Ordering::Relaxed));
    let r = as_bit_pattern(&V);
    let _copy = r[0]; // UB: non-atomic read races with atomic store on thread t
    t.join().unwrap();
}
```

**Comment by RalfJung** [comment]
> that's not actually true afaict, data race:

Ah right, it needs to be `UnsafeCell<MaybeUninit<u8>>` or something like that.

**Comment by K4rakara** [comment]
I *think* I might have found a bug ([playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ff81411341bc337b71b7f3d8bbab4d72)).

The following code:
```rust
#![feature(adt_const_params)]
#![feature(const_trait_impl)]
#![feature(generic_const_exprs)]

use std::marker::PhantomData;
use std::mem;

pub struct Assert<const T: bool> {
    __prevent_contstruction: PhantomData<()>,
}

impl const IsTrue for Assert<true> { }

#[const_trait]
pub trait IsTrue { }

pub struct Foo;

impl Foo {
    pub const fn bar(&self)
    where
        Assert<{ mem::size_of_val(self) == 0 }>: IsTrue,
    {
        todo!()
    }
}
```

Causes the compiler to emit the following error:

```
error[E0424]: expected value, found module `self`
  --> src/lib.rs:22:35
   |
20 | /     pub const fn bar(&self)
21 | |     where
22 | |         Assert<{ mem::size_of_val(self) == 0 }>: IsTrue,
   | |                                   ^^^^ `self` value is a keyword only available in methods with a `self` parameter
23 | |     {
24 | |         todo!()
25 | |     }
   | |_____- this function has a `self` parameter, but a macro invocation can only access identifiers it receives from parameters
```

Whether you think `self` parameters should be allowed in generic const exprs is probably a point of contention, but I definitely don't think this should be mentioning macros in its error message.

**Comment by oli-obk** [comment]
I am locking the issue as it is a tracking issue. Please open zulip threads for discussions and new issues for bug reports.

**Comment by RalfJung** [comment]
Before const generic of reference type get stabilized, we need to carefully think about and resolve [this interaction](https://github.com/rust-lang/rust/pull/119614#issuecomment-1880729454) with consts that point to statics. What is the best way to track this? Usually I'd add an item to the original issue, but here that seems to be delegated to a separate repo, but that repo seems to have fairly strict rules for what the issues look like and I really just want to leave a bullet point as a starting point for something to be explored eventually.

**Comment by BoxyUwU** [comment]
I would open an issue and label it `F-adt_const_params`

**Comment by RalfJung** [comment]
Ah right, this is even the wrong tracking issue *oops*

**Comment by darconeous** [comment]
It would be great to be able to enable very limited use cases, such as only for functions returning the generic. That allows you to have types that can build on each other incrementally.

[I made a quick example on the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=090a530247688b5d0307024fb0005acd). What is really nice about this is that you can construct really complex types statically, entirely at runtime.

This would make the micropython crate I'm working on quite a bit easier to maintain, and it seems like there aren't a lot of edge cases when using the feature in this way.

**Comment by comex** [comment]
See also the [min_generic_const_arguments](https://github.com/rust-lang/rust-project-goals/issues/100) project goal.

However, I have no idea whether this would support your use case.  The plan is apparently to stabilize some minimal subset of expressions in const generic arguments, but neither the project goal nor the GitHub issue so much as hints at what exactly that subset might contain.

**Comment by BoxyUwU** [comment]
@comex @darconeous the tracking issue for mgca (#132980) links to a document (as of a couple days ago),that explains the idea behind mgca, the project goal issue *used to* but it seems that the issue recently got re-edited back to its original state without all the additional information that had been added... oh well.. you can find the document about mgca [here](https://hackmd.io/OsVAAEgOS2ajDwxg44kN-A?view)

I dont expect mgca to be viable for what you're doing as it involves generic arithmetic which is *especially* hard to support

**Comment by dzmitry-lahoda** [comment]
used witness types with private constructors and const function in bounded vec impl to do exactly `generic_const_exprs` thing https://github.com/ergoplatform/bounded-vec/pull/31 if somebody needs const constraints on types now (and for last 4 years).

**Comment by dzmitry-lahoda** [comment]
used witness types with private constructors and const function in bounded vec impl to do exactly `generic_const_exprs` thing https://github.com/ergoplatform/bounded-vec/pull/31 if somebody needs const constraints on types now (and for last 4 years). UPDATE: updated - need to use const fn and const { block inside

**Comment by comex** [comment]
Sounds useful for some cases, but doesn't work if a type definition needs to actually depend on a constant expression (e.g. for an array length).

**Comment by oli-obk** [comment]
Please keep the discussion in dedicated issues or zulip threads

> A tracking issue is however not meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

**Comment by daniel-pfeiffer** [comment]
@lcnr In the overview the 1st generic is not closed.
